_posts/2020-10-24-text-layout.md

```diff
@@ -0,0 +1,217 @@
+
+I should note one thing, which might help people doing archaeology of legacy text stacks: it used to be pretty common for text layout to resolve "compatibility" forms such as NFKC and NFKD, and this can lead to various problems. But today it is more common to solve that particular problem by providing a font stack with *massive* Unicode coverage, including all the code points in the relevant compatibility ranges.
+
+### Script
+
+The *shaping* of text, or the transformation of a sequence of code points into a sequence of positioned glyphs, depends on the script. Some scripts, such as Arabic and Devanagari, have extremely elaborate shaping rules, while others, such as Chinese, are a fairly straightforward mapping from code point into glyph. Latin is somewhere in the middle, starting with a straightforward mapping, but ligatures and kerning are also required for high quality text layout.
+
+Determining script runs is reasonably straightforward - many characters have a Unicode script property which uniquely identifies which script they belong to. However, some characters, such as space, are "common," so the assigned script just continues the previous run.
+
+A simple example is "hello мир". This string is broken into two script runs: "hello " is Latn, and "мир" is Cyrl.
+
+### Shaping (cluster)
+
+At this point, we have a run of constant style, font, direction, and script. It is ready for *shaping.* Shaping is a complicated process that converts a string (sequence of Unicode code points) into positioned glyphs. For the purpose of this blog post, we can generally treat it as a black box. Fortunately, a very high quality open source implementation exists, in the form of HarfBuzz.
+
+We're not *quite* done with segmentation, though, as shaping assigns substrings in the input to [clusters] of glyphs. The correspondence depends a lot on the font. In Latin, the string "fi" is often shaped to a single glyph ( a ligature). For complex scripts such as Devanagari, a cluster is most often a syllable in the source text, and complex reordering can happen within the cluster.
+
+Clusters are important for *hit testing,* or determining the correspondence between a physical cursor position in the text layout and the offset within the text. Generally, they can be ignored if the text will only be rendered, not edited (or selected).
+
+Note that these shaping clusters are distinct from grapheme clusters. The "fi" example has two grapheme clusters but a single shaping cluster, so a grapheme cluster boundary can cut a shaping cluster. Since it's possible to move the cursor between the "f" and "i", one tricky problem is to determine the cursor location in that case. Fonts *do* have a [caret table], but implementation is spotty. A more robust solution is to portion the width of the cluster equally to each grapheme cluster within the cluster. See also [Let’s Stop Ascribing Meaning to Code Points] for a detailed dive into grapheme clusters.
```

Comment:
Related here is the good old Firefox [“f soft-hyphen f” that can splits a ligature glyph across lines](https://bugzilla.mozilla.org/show_bug.cgi?id=479829). And of course there’s no universally-ideal answer here, because the ligature probably changes the metrics and thus potentially the line break position. That’s one of my favourite still-extant bugs. (I haven’t delved deeply, but I think this sort of thing is part of what the harfbuzz#1463 and allsorts#29 issues are about.)

Reply:
Fantastic, I actually added a paragraph and the image from the big. The post is short on images illustrating the concepts, and I think I'll try to add some.